crates/core/src/store.rs

```diff
@@ -19,6 +20,28 @@ use crate::{
 #[cfg(doc)]
 use crate::EngineBuilder;
 
+/// Wrapper for the Preview 1 and Preview 2 versions of `WasiCtx`.
+///
+/// Currently, only WAGI uses Preview 1, while everything else uses Preview 2 (possibly via an adapter).  WAGI is
+/// stuck on Preview 1 and modules because there's no reliable way to wrap an arbitrary Preview 1 command in a
+/// component -- the Preview 1 -> 2 adapter only works with modules that either export `canonical_abi_realloc`
+/// (e.g. native Spin apps) or use a recent version of `wasi-sdk`, which contains patches to allow the adapter to
+/// safely allocate memory via `memory.grow`.
+///
+/// In theory, someone could build a WAGI app using a new-enough version of `wasi-sdk` and wrap it in a component
+/// using the adapter, but that wouldn't add any value beyond leaving it as a module, and any toolchain capable of
+/// natively producing components will be capable enough to produce native Spin apps, so we probably won't ever
+/// support WAGI components.
+///
+// TODO: As of this writing, the plan is to merge the WASI Preview 1 and Preview 2 implementations together, at
+// which point we'll be able to avoid all the duplication here and below.
+pub struct Wasi {
+    /// Preview 1 `WasiCtx`
+    pub preview1: wasi_preview1::WasiCtx,
+    /// Preview 2 `WasiCtx`
+    pub preview2: wasi_preview2::WasiCtx,
+}
```

Comment:
It would be nice if we could make this an enum instead, switched by a flag to `StoreBuilder::new`. Mostly so that we could bail if the "wrong" version is used with features below.

Do we have any sense of the timeline for merging the implementations?

Reply:
Yeah, I like the enum idea.  Not sure about the timeline for merging; I'll look into it.  It probably depends on when and how much we contribute to making it happen :)

Reply:
I'm implementing the enum idea now.  It will also allow me to get rid of the `Clone` bounds I added in various places, so definitely an improvement.